src/inmanta/ast/type.py

```diff
@@ -390,46 +239,61 @@ def validate(cls, value):
             return True
 
         if not isinstance(value, list):
-            raise RuntimeException(None, "Invalid value '%s', expected list" % value)
+            raise RuntimeException(None, "Invalid value '%s', expected %s" % (value, self.type_string()))
 
         return True
 
-    #     @classmethod
-    #     def __str__(cls):
-    #         return "list"
-
-    @classmethod
-    def type_string(cls):
+    def type_string(self) -> str:
         return "list"
 
-    def __str__(self):
-        return list.__str__(self)
+    def get_location(self) -> Location:
+        return None
+
 
-    @classmethod
-    def get_location(cls) -> Location:
+class TypedList(List):
+    def __init__(self, element_type: Type) -> None:
+        List.__init__(self)
+        self.element_type: Type = element_type
+
+    def normalize(self) -> None:
+        self.element_type.normalize()
+
+    def validate(self, value: Optional[object]) -> bool:
+        if not List.validate(self, value):
+            return False
+
+        assert isinstance(value, list)
+        for element in value:
+            self.element_type.validate(element)
+
+        return True
+
+    def type_string(self) -> str:
+        return "%s[]" % (self.element_type.type_string())
+
+    def get_location(self) -> Location:
         return None
 
 
-class Dict(Type, dict):
+class LiteralList(TypedList):
     """
-        This class represents a list type in the configuration model. (instances represent instances)
+    This class represents a list type containing only literals.
+    (instances of the class represent the type)
     """
 
-    def __init__(self):
-        Type.__init__(self)
-        dict.__init__(self)
+    def __init__(self) -> None:
+        TypedList.__init__(self, Literal())
 
-    @classmethod
-    def cast(cls, value):
-        """
-            Cast the given value to a string
 
-            :see CastableType#cast
-        """
-        return dict(value)
+class Dict(Type):
+    """
+        This class represents a list type in the configuration model. (instances represent instances)
+    """
 
-    @classmethod
-    def validate(cls, value):
+    def __init__(self) -> None:
+        Type.__init__(self)
+
+    def validate(self, value: Optional[object]) -> bool:
```

Comment:
The existing 

```
if value is None:
            return True	       
```

is a bit scary, but it is existing, so we can let irt be for now.